execute/tokendata/usdc/http.go

```diff
@@ -0,0 +1,198 @@
+package usdc
+
+import (
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/time/rate"
+)
+
+const (
+	apiVersion      = "v1"
+	attestationPath = "attestations"
+
+	// defaultCoolDownDurationSec defines the default time to wait after getting rate limited.
+	// this value is only used if the 429 response does not contain the Retry-After header
+	defaultCoolDownDuration = 5 * time.Minute
+
+	// maxCoolDownDuration defines the maximum duration we can wait till firing the next request
+	maxCoolDownDuration = 10 * time.Minute
+)
+
+type HTTPClient interface {
+	// Get calls the USDC attestation API with the given USDC message hash.
+	// The attestation service rate limit is 10 requests per second. If you exceed 10 requests
+	// per second, the service blocks all API requests for the next 5 minutes and returns an
+	// HTTP 429 response.
+	//
+	// Documentation:
+	//
+	//	https://developers.circle.com/stablecoins/reference/getattestation
+	//	https://developers.circle.com/stablecoins/docs/transfer-usdc-on-testnet-from-ethereum-to-avalanche
+	Get(ctx context.Context, messageHash [32]byte) ([]byte, int, error)
+}
+
+// httpClient is a client for the USDC attestation API. It encapsulates all the details specific to the Attestation API:
+// - rate limiting
+// - cool down period
+// - parsing JSON response and handling errors
+// Therefore AttestationClient is a higher level abstraction that uses httpClient to fetch attestations and can be more
+// oriented around caching/processing the attestation data instead of handling the API specifics.
+type httpClient struct {
+	apiURL     string
+	apiTimeout time.Duration
+	rate       *rate.Limiter
+	// coolDownUntil defines whether requests are blocked or not.
+	coolDownUntil time.Time
+	coolDownMu    *sync.RWMutex
+}
+
+//nolint:revive
+func NewHTTPClient(api string, apiInterval time.Duration, apiTimeout time.Duration) *httpClient {
+	return &httpClient{
+		apiURL:     api,
```

Comment:
q: Do we wanna do some kind of parsing on the passed in URL so that we don't construct garbage when doing the `fmt.Sprintf` below?

Reply:
yeah, probably makes sense to be defensive here